herbert_core/utilities/maths/noisify.m

```diff
@@ -17,40 +17,79 @@
 %           Add noise with Poisson distribution, where the mean value at
 %           a point is the value y.
 %
-%   If no input errors, e, just set e=[]
+%   >> [yout, eout] = noisify(y,e,[factor,]'maximum_value',maxval)
+%           Add noise with Gaussian distribution, calculating the standard
+%           deviation by an externally provided maximum y value.
+%           The max value is preceded by a keyword string 'maximum_value'.
+%           Typically this value will be the overall maximum value if the
+%           data is processed page by page and the maximum must be
+%           extracted before this processing.
+%           Omitting factor will use the default value 0.1.
+%
+%   If no input errors, e, just set e=[]. Note that eout will be created
+%   regardless of whether e is present, so the first overload output is
+%   probably incomplete.
+
+    % deal with optional arguments
+    fac = 0.1
+    is_poisson = false;
+    ymax_calculated = false;
+    if nargin>=2 % avoids the case where the only paramemter is y; for >=2 e will not be optional
+        [fac, is_poisson, ymax] = parse_args(y, e, fac, varargin{:});
+        ymax_calculated = true;
+    end
 
-if nargin==3 && ischar(varargin{1})
-    if is_stringmatchi(varargin{1},'poisson')
+    % Use Poisson distribution and ignore other arguments
+    if is_poisson
         yout=zeros(size(y));
         for i=1:numel(y)
             yout(i)=randpoisson(abs(y(i)));
         end
-        eout=abs(y);  % the inpuy y is the mean and variance of the Poisson distribution
+        eout=abs(y);  % the input y is the mean and variance of the Poisson distribution
     else
-        error('Unrecognised random sampling distribution')
+        if ~ymax_calculated
+            ymax = max(abs(y(:)))
```

Comment:
this logic is repeated in the `parse_args` function. I see who it is needed here for `nargin == 1`. 

Can the `parse_args` function be adapted to handle that case too so the `ymax_calculated` flag and repeated logic are not required?

Reply:
Undoubtedly, but this is working now, and the next work with noisify in Horace depends on it, so that improvement will happen later.